### PR TITLE
Restrict the size of the permissions metadata store

### LINKED ISSUE #8569: Site metadata has unbounded growth
Site metadata, collected for permissions purposes, is currently stored in the background state and never trimmed or bounded. We should consider how to bound the size of this store.

### PR DESCRIPTION
**Pull request by rekmarks**
Closes #8569

This PR bounds the size of the permissions metadata store to the number of domains with permissions.
- Metadata is _never_ deleted for domains with permissions
  - This should be fine in the short to medium term
  - This may be untenable in the long term, but we need to make some fundamental improvements to our JSON RPC implementation before we can address it in version `8.x`
- When rehydrating the store on boot, we delete the metadata of all domains without permissions, if any
- At runtime, we keep track of newly added metadata origins via the `_pendingSiteMetadata` set
  - If adding a new origin would ever grow this set beyond 100 in size, we delete the oldest entry in it, and _if that oldest origin has no permissions_, we also delete its metadata from the store
- We do _not_ delete metadata when permissions are removed, to prevent a degraded user experience
  - We still delete the metadata of such domains on boot, but not otherwise
  - This will be addressed when we can request permissions on demand in `8.x`

### CODE REVIEW

File: app/scripts/controllers/permissions/index.js [hunk 1]
```diff
@@ -510,6 +512,75 @@ export class PermissionsController {
     })
   }
 
+  /**
+   * Removes all known domains and their related permissions.
+   */
+  clearPermissions () {
```
Comment (rekmarks) [inline app/scripts/controllers/permissions/index.js:518]: This function was moved, because its placement didn't make any sense. It hasn't been changed.

File: app/scripts/controllers/permissions/index.js [hunk 2]
```diff
@@ -510,6 +512,75 @@ export class PermissionsController {
...
+    this.permissions.clearDomains()
+    this.notifyAllDomains({
+      method: NOTIFICATION_NAMES.accountsChanged,
+      result: [],
+    })
+  }
```
Comment (rekmarks) [inline app/scripts/controllers/permissions/index.js:524]: This function was moved, because its placement didn't make any sense. It hasn't been changed.

**Comment by Gudahtt** [comment]
Do we ever want to trim metadata for sites _with_ permissions? 🤔 I don't see the harm in keeping that. We use it in the UI anyway - we'd have to remove the permission too to avoid weird UI bugs.

File: app/scripts/controllers/permissions/index.js [hunk 3]
```diff
@@ -510,6 +507,119 @@ export class PermissionsController {
...
+      lastUpdated: Date.now(),
+    }
+    this._newMetadataOrigins.add(origin)
+
+    this._setDomainMetadata(newMetadataState)
+  }
+
+  /**
+   * Deletes metadata for domains without permissions when the metadata store
+   * has grown too large. Highly dependent on logic of addDomainMetadata.
```
Comment (Gudahtt) [inline app/scripts/controllers/permissions/index.js:552]: >Highly dependent on logic of addDomainMetadata

Could this be inlined in `addDomainMetadata` instead then? The behavior of this function is.... pretty strange. I don't think it makes sense as an independent function.

File: app/scripts/controllers/permissions/index.js [hunk 4]
```diff
@@ -510,6 +507,119 @@ export class PermissionsController {
...
+    if (Object.keys(newMetadataState).length >= METADATA_STORE_MAX_SIZE) {
+      this._popDomainMetadata(newMetadataState)
+    }
+
+    newMetadataState[origin] = {
+      ...oldMetadataState[origin],
+      ...metadata,
+      lastUpdated: Date.now(),
+    }
+    this._newMetadataOrigins.add(origin)
```
Comment (Gudahtt) [inline app/scripts/controllers/permissions/index.js:545]: I'm having trouble following why this `Set` is here. Why is this useful?

File: app/scripts/controllers/permissions/index.js [hunk 5]
```diff
@@ -510,6 +507,119 @@ export class PermissionsController {
...
+   *
+   * @param {string} origin - The origin whose domain metadata to store.
+   * @param {Object} metadata - The domain's metadata that will be stored.
+   */
+  addDomainMetadata (origin, metadata) {
+
+    const oldMetadataState = this.store.getState()[METADATA_STORE_KEY]
+    const newMetadataState = { ...oldMetadataState }
+
+    if (Object.keys(newMetadataState).length >= METADATA_STORE_MAX_SIZE) {
```
Comment (Gudahtt) [inline app/scripts/controllers/permissions/index.js:536]: We should only be considering the length of the metadata for sites without permissions. Otherwise things will get pretty weird for users with >100 sites connected - they'll be unable to gather more metadata, so our UI is gonna look pretty broken.

**Review by Gudahtt — CHANGES_REQUESTED** [review]

File: app/scripts/controllers/permissions/index.js (hunk 5 above)
Comment (rekmarks) [reply app/scripts/controllers/permissions/index.js:537]: What this function does should be clearer now that `popDomainMetadata` is inlined. 

We consider the size of the entire store, but only delete metadata for domains without permissions. Which domains have been added during this session is kept track of in the `_newMetadataOrigins` set.

We iterate over the values of `_newMetadataOrigins` and delete the metadata of the first domain without metadata that we find. We _also_ delete all origins we iterate over from `_newMetadataOrigins` because they will either be deleted entirely, or they have permissions and won't be deleted this session.

File: app/scripts/controllers/permissions/index.js (hunk 4 above)
Comment (rekmarks) [reply app/scripts/controllers/permissions/index.js:557]: See https://github.com/MetaMask/metamask-extension/pull/8596/files#r427416153

File: app/scripts/controllers/permissions/index.js (hunk 3 above)
Comment (rekmarks) [reply app/scripts/controllers/permissions/index.js]: Done in faaa60b

File: app/scripts/controllers/permissions/index.js (hunk 5 above)
Comment (Gudahtt) [reply app/scripts/controllers/permissions/index.js:538]: I just don't follow why "domains that have been added during this session" is something we need to keep track of. What is a "session" in this contenxt, and why would we want to consider only these domains?

Comment (Gudahtt) [reply app/scripts/controllers/permissions/index.js:538]: The way I'd imagine we'd remove site metadata is to iterate over the site metadata, and remove the first one we find without permissions (as the iteration order should be guaranteed to be the same as the insertion order as of ES2015, so the first one will also be the oldest). Then we wouldn't need to track this extra state.

Comment (Gudahtt) [reply app/scripts/controllers/permissions/index.js:538]: Oh, but to return to the original comment in this thread:

>We consider the size of the entire store

This is the part that I think is broken; if the user has >=100 sites connected, new sites won't correctly populate the site metadata. We could fix this by only considering metadata for sites without permissions in the first place.

Comment (rekmarks) [reply app/scripts/controllers/permissions/index.js:538]: > What is a "session" in this context, and why would we want to consider only these domains?

By "session", I mean one lifecycle of the MetaMask background process. So, from the time the extension background process starts until it exits.

> The way I'd imagine we'd remove site metadata is to iterate over the site metadata, and remove the first one we find without permissions

> This is the part that I think is broken; if the user has >=100 sites connected, new sites won't correctly populate the site metadata. We could fix this by only considering metadata for sites without permissions in the first place.

That's what this implementation currently does, except that it counts all domains with metadata, as opposed to only those without permissions, and it does not consider domains that ever had metadata.

I can change the trigger to only count domains without metadata, or rather the size of the `_newMetadataOrigins`.

The reason I'm not just iterating over the entire store and deleting the first thing I find is (1) I think it's inelegant (and potentially inefficient) to iterate over the entire store every time we want to delete something, and (2) we can't safely delete metadata for domains that have had permissions during a session of MetaMask, until we can request metadata on demand.

For (2), consider a user that removes permissions from a connected site, and the site's metadata gets deleted. Now the user clicks connect on that site (or opens some other tab expecting metadata), without refreshing its tab. There's now no metadata for that site, and our UI is broken.

Given all of the above, I added `_newMetadataOrigins` to keep track of domains added during the current session of MetaMask, and only delete metadata for domains without metadata in that set, in insertion order (FIFO).

I'll update the condition triggering the deletion attempt to check the size of the set instead of the size of the metadata store. That way, we should always have the metadata we need, although the UI may still break if a user leaves a tab open for a very long time and attempts to connect after opening 100 additional sites.

File: app/scripts/controllers/permissions/index.js [hunk 6]
```diff
@@ -35,8 +36,8 @@ export class PermissionsController {
     restoredPermissions = {},
     restoredState = {}) {
 
+    // more state set further down
```
Comment (whymarrh) [inline app/scripts/controllers/permissions/index.js:39]: Nit: this comment isn't particularly useful

File: app/scripts/controllers/permissions/index.js [hunk 7]
```diff
@@ -61,6 +62,9 @@ export class PermissionsController {
     this._lastSelectedAddress = preferences.getState().selectedAddress
     this.preferences = preferences
 
+    // _initializeMetadataStore requires _initializePermissions
```
Comment (whymarrh) [inline app/scripts/controllers/permissions/index.js:65]: This comment should be in the docs for the method, the method JSDoc

File: app/scripts/controllers/permissions/index.js [hunk 8]
```diff
@@ -510,6 +507,100 @@ export class PermissionsController {
...
+    this.notifyAllDomains({
+      method: NOTIFICATION_NAMES.accountsChanged,
+      result: [],
+    })
+  }
+
+  /**
+   * Stores domain metadata for the given origin (domain).
+   * Deletes metadata for domains without permissions in a FIFO manner.
+   * Metadata is only deleted for domains with permissions on boot in order to
```
Comment (whymarrh) [inline app/scripts/controllers/permissions/index.js:524]: Is this supposed to be _without_?

```suggestion
   * Metadata is only deleted for domains without permissions on boot in order to
```

File: app/scripts/controllers/permissions/index.js [hunk 9]
```diff
@@ -510,6 +507,100 @@ export class PermissionsController {
...
+    this._setDomainMetadata(newMetadataState)
+  }
+
+  /**
+   * Trims the given metadataState object by removing metadata for all origins
+   * without permissions.
+   * Returns a new object; does not mutate the argument.
+   *
+   * @param {Object} metadataState - The metadata store state object to trim.
+   * @returns {Object} The mutated metadata state object.
```
Comment (whymarrh) [inline app/scripts/controllers/permissions/index.js:579]: I think this might be outdated?

```suggestion
   * @returns {Object} The new metadata state object.
```

**Review by whymarrh — CHANGES_REQUESTED** [review]
> - If there are 100 or more domains in the metadata store, we attempt to remove a domain without permissions before adding a new one
>   - This only fails if there is no domain metadata for a domain without permissions

What is supposed to happen in this failure case?

File: app/scripts/controllers/permissions/index.js [hunk 10]
```diff
@@ -510,6 +507,100 @@ export class PermissionsController {
...
+      result: [],
+    })
+  }
+
+  /**
+   * Stores domain metadata for the given origin (domain).
+   * Deletes metadata for domains without permissions in a FIFO manner.
+   * Metadata is only deleted for domains with permissions on boot in order to
+   * prevent a degraded user experience, as metadata currently can't be
+   * requested on-demand.
```
Comment (Gudahtt) [inline app/scripts/controllers/permissions/index.js:526]: Nit: I'm having trouble understanding this description, and I'm not sure it's all true anymore?

Something like this would make more sense to me:

```suggestion
   * If the pending site metadata exceeds the max cache size, the oldest metadata
   * will be deletedto make room.
```

This makes it more clear _when_ and _why_ metadata is deleted, and omits the detail about what happens on boot, which doesn't seem relevant to this function.

File: app/scripts/controllers/permissions/index.js [hunk 11]
```diff
@@ -510,6 +507,100 @@ export class PermissionsController {
...
+    const newMetadataState = { ...oldMetadataState }
+
+    // delete pending metadata origin from queue, and delete its metadata if
+    // it doesn't have any permissions
+    if (this._pendingSiteMetadata.size >= METADATA_CACHE_MAX_SIZE) {
+      const permissionsDomains = this.permissions.getDomains()
+
+      const oldOrigin = this._pendingSiteMetadata.values().next().value
+      this._pendingSiteMetadata.delete(oldOrigin)
+      if (!permissionsDomains[oldOrigin]) {
```
Comment (Gudahtt) [inline app/scripts/controllers/permissions/index.js:543]: Is this check still needed? I thought that `_pendingSiteMetadata` now represented the set of site metadata that hadn't been given permissions yet, so this would always be true.

File: app/scripts/controllers/permissions/index.js [hunk 12]
```diff
@@ -510,6 +507,100 @@ export class PermissionsController {
...
+   * @param {Object} metadata - The domain's metadata that will be stored.
+   */
+  addDomainMetadata (origin, metadata) {
+
+    const oldMetadataState = this.store.getState()[METADATA_STORE_KEY]
+    const newMetadataState = { ...oldMetadataState }
+
+    // delete pending metadata origin from queue, and delete its metadata if
+    // it doesn't have any permissions
+    if (this._pendingSiteMetadata.size >= METADATA_CACHE_MAX_SIZE) {
```
Comment (Gudahtt) [inline app/scripts/controllers/permissions/index.js:538]: This should first check to see whether the metadata for this site is already stored. If it is, all of this can be skipped.

The way it's implemented now, we might be deleting and then re-adding the same thing each page load (if it happens to be the oldest thing in the cache anyway).

File: app/scripts/controllers/permissions/index.js (hunk 11 above)
Comment (Gudahtt) [reply app/scripts/controllers/permissions/index.js:541]: Oh I see - this is the only place they're removed.

**Comment by rekmarks** [comment]
@whymarrh all inline comments addressed in d3c7d89.

Regarding your question:

> What is supposed to happen in this failure case?

The PR description was outdated; there is no failure case. The bullet point you referred to has been updated to read:

- At runtime, we keep track of newly added metadata origins via the _pendingSiteMetadata set
  - If adding a new origin would ever grow this set beyond 100 in size, we delete the oldest entry in it, and if that oldest origin has no permissions, we also delete its metadata from the store

**Comment by rekmarks** [comment]
@Gudahtt I missed your review before responding to @whymarrh, will comment on your requested changes in turn.

File: app/scripts/controllers/permissions/index.js (hunk 10 above)
Comment (rekmarks) [reply app/scripts/controllers/permissions/index.js]: I updated this comment in d3c7d89, please let me know if it suffices.

File: app/scripts/controllers/permissions/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (rekmarks) [reply app/scripts/controllers/permissions/index.js]: Added something very close to this in c3b7128

File: app/scripts/controllers/permissions/index.js (hunk 10 above)
Comment (Gudahtt) [reply app/scripts/controllers/permissions/index.js]: Seems fine to me!

File: app/scripts/controllers/permissions/index.js (hunk 12 above)
Comment (rekmarks) [reply app/scripts/controllers/permissions/index.js:537]: Well, deleting and re-adding origins from the `_pendingSiteMetadata` set is a feature, not a bug, because we want the insertion order to be updated.

We could end up deleting and re-adding the `siteMetadata`, too, but that'll have to get overwritten anyway (it may have changed), and `delete` is cheap.

Comment (Gudahtt) [reply app/scripts/controllers/permissions/index.js:537]: Hmm. OK. Well that's not ideal from a performance standpoint, but I guess it's negligible next to our other performance issues! I suppose this is OK for now.

File: app/scripts/controllers/permissions/index.js (hunk 5 above)
Comment (rekmarks) [reply app/scripts/controllers/permissions/index.js]: This was resolved out of band.

**Review by Gudahtt — DISMISSED** [review]
LGTM!

File: test/unit/app/controllers/permissions/permissions-controller-test.js [hunk 13]
```diff
@@ -1318,6 +1324,180 @@ describe('permissions controller', function () {
...
+          },
+        },
+      })
+
+      permController.addDomainMetadata(ORIGINS.b, { foo: 'bar' })
+
+      assert.ok(
+        permController.store.getState.called,
+        'should have called store.getState'
+      )
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply test/unit/app/controllers/permissions/permissions-controller-test.js:1365]: In general I agree, but here the implementation of the test hinges on the return value of `getState` (see the initialization of the `fake` ~10 lines above).

Obviously, the test should fail if the implementation stops relying on `getState`, but it doesn't seem completely pointless to me in this case.

[the remark this answers is not in the document]
Comment (Gudahtt) [reply test/unit/app/controllers/permissions/permissions-controller-test.js:1365]: Ah I see - that makes more sense. I missed that.

**Review by Gudahtt — APPROVED** [review]

**Review by whymarrh — APPROVED** [review]
:shipit:
